{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "68ebfeea-d2a7-4e93-9b7d-0a1693ace40b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import gzip\n",
    "import concurrent.futures\n",
    "\n",
    "from collections import defaultdict\n",
    "from urllib.request import urlopen\n",
    "'''\n",
    "To run extraction on your computer you first need to download the files from:\n",
    "https://hgdownload.soe.ucsc.edu/goldenPath/hg38/multiz100way/.\n",
    "\n",
    "Run the comma×žds:\n",
    "mut_extractor = MPMutExtractor(folder, file_names, species, url)\n",
    "mut_extractor.get_mutation_counts()\n",
    "\n",
    "When the parameters are:\n",
    "folder: The folder containing the downloaded maf files with the different chromosomes alignment data\n",
    "file_names: The name of all the files with the data from the folder\n",
    "species: A String array with the three (scientific) names of the species you want to analyse.\n",
    "example = [\"Ovis aries\", \"Capra hircus\", \"Bos taurus\"], where Bos taurus is the outgroup\n",
    "url - no need to specify, but should contain the url of the multiz100way in UCSC\n",
    "'''\n",
    "\n",
    "\n",
    "\n",
    "# Multiprocessing mutational count extractor\n",
    "class MPMutExtractor:\n",
    "    \"\"\"\n",
    "    This class uses multiprocessing to extract mutation counts\n",
    "    It extracts counts from multiple chromosome files concurrently using Chromosome_MutSig class.\n",
    "    \n",
    "    Important Attributes:\n",
    "    species_names (String array): scientific name of the 3 species to be used (species[2]) is the outlgroup.\n",
    "    folder (String): The path of the folder containing alignment files\n",
    "    file_names (String array): The file names that the sequencing files are in.\n",
    "    \"\"\"\n",
    "    \n",
    "\n",
    "    def __init__(self, \n",
    "                 folder = r\"C:\\Users\\KerenYlab\\Downloads\" + \"\\\\\", \n",
    "                 file_names = ['chr1.maf.gz'],\n",
    "                 species = [\"Ovis aries\", \"Capra hircus\", \"Bos taurus\"],\n",
    "                 url = \"https://hgdownload.soe.ucsc.edu/goldenPath/hg38/multiz100way/\"):\n",
    "        \n",
    "        \"\"\"\n",
    "        A constructor to initialize a new MPMutExtractor instance.\n",
    "        \n",
    "        Parameters:\n",
    "        self.html : html page information with about maf alignment files\n",
    "        folder, file_names, species_names : as described above\n",
    "        \"\"\"\n",
    "        \n",
    "        \n",
    "        page = urlopen(url)\n",
    "        html_bytes = page.read()\n",
    "        self.html = html_bytes.decode(\"utf-8\")\n",
    "        self.folder = folder\n",
    "        self.file_names = file_names\n",
    "        self.species_names = species\n",
    "    \n",
    "    \n",
    "    def get_mutation_counts(self):\n",
    "        \"\"\"\n",
    "        This function that runs the concurrent mutation count.\n",
    "\n",
    "        Returns:\n",
    "        a list of the mutation count for each chromosome (each mutation count is a dict)\n",
    "        \"\"\"\n",
    "        \n",
    "        self.__get_species_alignment_names(self.species_names)     \n",
    "        results = None\n",
    "        \n",
    "        # Create an Executor object that runs the processes\n",
    "        with concurrent.futures.ThreadPoolExecutor() as executor:\n",
    "            results = executor.map(self.__get_chromosome_mutational_count, self.file_names)\n",
    "            self.mutation_counts = list(results)\n",
    "        return self.mutation_counts\n",
    "            \n",
    "        \n",
    "    # The function that runs the mutation count on a single chromosome\n",
    "    # Returns the mutation count from the chromosome (a dict)\n",
    "    def __get_chromosome_mutational_count(self, file_name):\n",
    "        chromosome_mutext = Chromosome_MutExt(self.folder, file_name, self.chosen_species)\n",
    "        mutational_counts = chromosome_mutext.get_mutation_counts()\n",
    "        return mutational_counts\n",
    "    \n",
    "        \n",
    "    # finds common names\n",
    "    def __find_common_names(self):\n",
    "        start = re.search(\"Assemblies used in these alignments:\", self.html)\n",
    "        end = re.search(\"---------------------------------------------------------------\", self.html)\n",
    "        text = self.html[start.end():end.start()]\n",
    "\n",
    "        names = []\n",
    "        for line in text.splitlines():\n",
    "            if (line != '') and not line.isspace() and line[0] != '=':\n",
    "                names.append(line[25:56].strip(' '))\n",
    "        common_names = '\\n'.join(names)\n",
    "        self.common_names = common_names\n",
    "        return common_names\n",
    "\n",
    "\n",
    "    #finds names for alignment\n",
    "    def __find_alignment_names(self):\n",
    "        names = []\n",
    "        start = re.search(\"Assemblies used in these alignments:\", self.html)\n",
    "        end = re.search(\"---------------------------------------------------------------\", self.html)\n",
    "        text = self.html[start.end():end.start()]\n",
    "        for line in text.splitlines():\n",
    "            if (line == '') or line.isspace() or line[0] == '=':\n",
    "                continue\n",
    "            names.append(re.findall(\"/\\S*\\s\", line)[-1][1:-1])\n",
    "        alignment_names = '\\n'.join(names)\n",
    "        self.alignment_names = alignment_names\n",
    "        return alignment_names\n",
    "    \n",
    "    #gets the alignment names for the 3 species given to it\n",
    "    def __get_species_alignment_names(self, species = [\"Ovis aries\", \"Capra hircus\", \"Bos taurus\"]):\n",
    "        \n",
    "        if not hasattr(self, \"common_names\") or not hasattr(self, \"alignment_names\"):\n",
    "            self.common_names = self.__find_common_names()\n",
    "            self.alignment_names = self.__find_alignment_names()\n",
    "\n",
    "        common_names = self.common_names.split('\\n')\n",
    "        alignment_names = self.alignment_names.split('\\n')\n",
    "        \n",
    "        chosen_species = [None]*3\n",
    "        \n",
    "        # 3 chosen taxa = Ovis aries, Capra hircus and outgroup Bos taurus\n",
    "        chosen_species[0] = np.array(alignment_names)[(np.array(common_names) == species[0])][0]\n",
    "        chosen_species[1] = np.array(alignment_names)[(np.array(common_names) == species[1])][0]\n",
    "        chosen_species[2] = np.array(alignment_names)[(np.array(common_names) == species[2])][0]\n",
    "        self.chosen_species = chosen_species\n",
    "        return chosen_species\n",
    "    \n",
    "    \n",
    "\n",
    "#sigle chromosome mutation counts extractor\n",
    "class Chromosome_MutExt: \n",
    "    \"\"\"\n",
    "    This class extracts in practice the mutation counts from a given chromosome\n",
    "    \n",
    "    Important Attributes:\n",
    "    chosen_species (String array): scientific name of the 3 species to be used (species[2]) is the outgroup.\n",
    "    folder (String): The path of the folder containing alignment file\n",
    "    file_name (String): The file name that the sequencing data is in.\n",
    "    mutation_count(dict): contains the mutation count extracted from the file\n",
    "    \"\"\"\n",
    "    def __init__(self,\n",
    "                 folder = r\"C:\\Users\\KerenYlab\\Downloads\" + \"\\\\\", \n",
    "                 file_name = 'chr1.maf.gz',\n",
    "                 species = ['oviAri3', 'capHir1', 'bosTau8']):\n",
    "        \n",
    "        \n",
    "        self.chosen_species = species\n",
    "        self.file_name = file_name\n",
    "        self.folder = folder\n",
    "        self.species1_mutation_count = defaultdict(int)\n",
    "        self.species2_mutation_count = defaultdict(int)\n",
    "        \n",
    "        \n",
    "    def get_mutation_counts(self):\n",
    "        \n",
    "        '''\n",
    "        Parses the file line by line to avoid reading too much into memory\n",
    "        Divides the lines into chunks (according to alignment file structure),\n",
    "        And processes each chunk using __parse_paragraph\n",
    "\n",
    "        Returns:\n",
    "            The final count from the whole file\n",
    "        '''\n",
    "        \n",
    "        species = [\"s \" + s for s in self.chosen_species]\n",
    "        self.species1_mutation_count = defaultdict(int)\n",
    "        self.species2_mutation_count = defaultdict(int)\n",
    "\n",
    "        with gzip.open(self.folder + self.file_name, 'rt', encoding='utf-8') as file:\n",
    "            chunk = [file.readline()]\n",
    "            for line in file:\n",
    "                if line[0] != 'a': \n",
    "                    chunk.append(line)\n",
    "                else: # beggining of new chunk\n",
    "                    chunk = ''.join(chunk)\n",
    "                    if species[0] in chunk and species[1] in chunk and species[2] in chunk:\n",
    "                        self.__parse_paragraph(chunk, species)\n",
    "                    chunk = [line]\n",
    "        # merges the 192 categories to 96\n",
    "        self.species1_final_mutation_count = self.__get_final_count(self.species1_mutation_count)\n",
    "        self.species2_final_mutation_count = self.__get_final_count(self.species2_mutation_count)\n",
    "        return (self.species1_final_mutation_count, self.species2_final_mutation_count)\n",
    "                       \n",
    "                    \n",
    "    anti_pattern = r\"[^ATGC]\"\n",
    "    name_regex = re.compile(r\"s \\w*.\")\n",
    "    sequence_regex = re.compile(r\"[atgcnATGCN-]+$\")\n",
    "    \n",
    "    def __parse_paragraph(self, paragraph, species):\n",
    "        '''\n",
    "        Parses a single chunk containing sequencing data from the 3 species (and others).\n",
    "        Extracts from it the mutation count and adds it to mutation_count dict\n",
    "        '''\n",
    "        sequences = {}\n",
    "        for line in paragraph.splitlines():\n",
    "            species_name = self.name_regex.findall(line)\n",
    "            if species_name and species_name[0][:-1] in species:\n",
    "                sequence = self.sequence_regex.findall(line)\n",
    "                if sequence:\n",
    "                    sequences[species_name[0][:-1]] = sequence[0].upper()\n",
    "                else:\n",
    "                    print(\"failed:\")\n",
    "                    print(line)\n",
    "                    break;\n",
    "        return self.__count_mutations(sequences, species)\n",
    "\n",
    "    \n",
    "\n",
    "    def __count_mutations(self, sequences, species):\n",
    "        '''\n",
    "        Counts from the 3 sequences all the mutations that can be detected\n",
    "        Counts only mutations between the two sister taxas, in which the outlies has\n",
    "        a nucleotide that is similar to that of one of the taxas.\n",
    "        Adds the counts to mutation_count\n",
    "        '''\n",
    "        \n",
    "        species1_mutation_count = self.species1_mutation_count\n",
    "        species2_mutation_count = self.species2_mutation_count\n",
    "        df = pd.DataFrame(columns = species)\n",
    "        df[species[0]] = list(sequences[species[0]])\n",
    "        df[species[1]] = list(sequences[species[1]])\n",
    "        df[species[2]] = list(sequences[species[2]])\n",
    "        df[species[0] + \" neighbors\"] = self.__get_neighbors(sequences[species[0]])\n",
    "        df[species[1] + \" neighbors\"] = self.__get_neighbors(sequences[species[1]])\n",
    "        df[species[2] + \" neighbors\"] = self.__get_neighbors(sequences[species[2]])\n",
    "        # creates a boolean column of all point mutation between sister taxa (the nucleotide is different and is not a gap)\n",
    "        # includes only mutations with adjacent nucleotides for all 3 taxa\n",
    "        df[\"mutations\"] = ((df[species[0]] != df[species[1]]) & \n",
    "                           (df[species[0] + \" neighbors\"] != '-') & \n",
    "                           (df[species[1] + \" neighbors\"] != '-') & \n",
    "                           (df[species[2] + \" neighbors\"] != '-'))\n",
    "\n",
    "        # creates a boolean column with only the mutation that happend in one taxa and not in the other\n",
    "        # this is important for establishing the type of mutation that occurred\n",
    "        # df[species[0] + \" neighbors\"].iloc[df[\"mutations\"] & ((df[species[0]] == df[species[2]])]\n",
    "\n",
    "        mutations = df.loc[df[\"mutations\"] & (df[species[0]] == df[species[2]])]\n",
    "        for mut in mutations.index:\n",
    "            mutation = mutations.loc[mut][species[1] + \" neighbors\"]\n",
    "            mutation = mutation[0] + mutations.loc[mut][species[2]] + mutation[2] + \"->\" + mutation[1]\n",
    "            species1_mutation_count[mutation] += 1\n",
    "\n",
    "        mutations = df.loc[df[\"mutations\"] & (df[species[1]] == df[species[2]])]\n",
    "        for mut in mutations.index:\n",
    "            mutation = mutations.loc[mut][species[0] + \" neighbors\"]\n",
    "            mutation = mutation[0] + mutations.loc[mut][species[2]] + mutation[2] + \"->\" + mutation[1]\n",
    "            species2_mutation_count[mutation] += 1\n",
    "\n",
    "        self.species1_mutation_count = species1_mutation_count\n",
    "        self.species2_mutation_count = species2_mutation_count\n",
    "        return df\n",
    "\n",
    "    #Calculated for each position the it's neighboring nucleotides (ignores '-' in alignment)\n",
    "    def __get_neighbors(self, string):\n",
    "        nucleotides = {'A','T','G','C'}\n",
    "        sequence = re.sub(self.anti_pattern, '', string)\n",
    "        sequence_len = len(sequence)\n",
    "        i = 0\n",
    "        neighbors = []\n",
    "        for idx, char in enumerate(string):\n",
    "            if i == sequence_len-1 or i == 0:  \n",
    "                neighbors.append('-')\n",
    "                if char in nucleotides:\n",
    "                    i+=1\n",
    "            elif char not in nucleotides:\n",
    "                neighbors.append('-')\n",
    "            else:\n",
    "                neighbors.append(sequence[i-1:i+2])\n",
    "                i += 1\n",
    "        return neighbors\n",
    "\n",
    "    \n",
    "    complement = {'A':'T',\n",
    "                  'T':'A',\n",
    "                  'C':'G',\n",
    "                  'G':'C'}\n",
    "    \n",
    "    # merges the mutations properly to reach 96 mutational categories\n",
    "    def __get_complement(self, mutation):\n",
    "        comp_mutation = ''.join([self.complement[nuc] for nuc in mutation[2::-1]])\n",
    "        comp_mutation += \"->\" + self.complement[mutation[-1]]\n",
    "        return comp_mutation\n",
    "\n",
    "    def __get_final_count(self, mutation_count):\n",
    "        merged_categories = defaultdict(int)\n",
    "\n",
    "        for mutation, count in mutation_count.items():\n",
    "            if mutation[1] in {'A', 'C'}:\n",
    "                merged_categories[self.__get_complement(mutation)] += count\n",
    "            else:\n",
    "                merged_categories[mutation] += count\n",
    "        return merged_categories\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f041213f-ca52-4501-aec6-9396fb658c39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "91\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "92"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cm = Chromosome_MutExt()\n",
    "cm.get_mutation_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
