{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7f8b9c5d-85f3-4677-9e7e-82a91bccb7c7",
   "metadata": {},
   "source": [
    "Class that deals with the extraction details and is helpful here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fadc5c7f-1c20-4631-bc4c-f707f6a56b5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#multiprocessing over one species triplet and many chromosomes\n",
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import gzip\n",
    "import os\n",
    "import sys\n",
    "import json\n",
    "\n",
    "import multiprocessing\n",
    "\n",
    "from collections import defaultdict\n",
    "from urllib.request import urlopen\n",
    "from ChromosomeMutExtractor import get_mutation_counts\n",
    "'''\n",
    "To run extraction on your computer you first need to download the files from:\n",
    "https://hgdownload.soe.ucsc.edu/goldenPath/hg38/multiz100way/.\n",
    "\n",
    "Run the commads:\n",
    "mut_extractor = MPMutExtractor(folder, file_names, species, url)\n",
    "mut_extractor.get_mutation_counts()\n",
    "\n",
    "Where:\n",
    "folder: The folder containing the downloaded maf files with the different chromosomes alignment data\n",
    "file_names: The name of all the files with the data from the folder\n",
    "species: A String array with the three (scientific) names of the species you want to analyse.\n",
    "example = [\"Ovis aries\", \"Capra hircus\", \"Bos taurus\"], where Bos taurus is the outgroup\n",
    "url - no need to specify, but should contain the url of the multiz100way in UCSC\n",
    "'''\n",
    "\n",
    "\n",
    "\n",
    "# Multiprocessing mutational count extractor\n",
    "class MPMutExtractor:\n",
    "    \"\"\"\n",
    "    This class uses multiprocessing to extract mutation counts\n",
    "    It extracts counts from multiple chromosome files concurrently using Chromosome_MutSig class.\n",
    "    \n",
    "    Important Attributes:\n",
    "    species_names (String array): scientific name of the 3 species to be used (species[2]) is the outlgroup.\n",
    "    folder (String): The path of the folder containing alignment files\n",
    "    file_names (String array): The file names that the sequencing files are in.\n",
    "    \"\"\"\n",
    "    \n",
    "\n",
    "    def __init__(self, \n",
    "                 folder = r\"C:\\Users\\KerenYlab\\Downloads\" + \"\\\\\", \n",
    "                 chromosomes_folder = \"chr_files\",\n",
    "                 species = [\"Ovis aries\", \"Capra hircus\", \"Bos taurus\"],\n",
    "                 url = \"https://hgdownload.soe.ucsc.edu/goldenPath/hg38/multiz100way/\"):\n",
    "        \n",
    "        \"\"\"\n",
    "        A constructor to initialize a new MPMutExtractor instance.\n",
    "        \n",
    "        Parameters:\n",
    "        self.html : html page information with about maf alignment files\n",
    "        folder, file_names, species_names : as described above\n",
    "        \"\"\"\n",
    "        \n",
    "        \n",
    "        page = urlopen(url)\n",
    "        html_bytes = page.read()\n",
    "        self.html = html_bytes.decode(\"utf-8\")\n",
    "        self.folder = os.path.join(os.getcwd(), chromosomes_folder)\n",
    "        self.chromosomes = [file.split('.')[0] for file in os.listdir(chromosomes_folder) if file.startswith(\"chr\")]\n",
    "        chromosome_files = [os.path.join(self.folder, file) for file in os.listdir(chromosomes_folder) if file.startswith(\"chr\")]\n",
    "        self.file_names = chromosome_files\n",
    "        self.species_names = species\n",
    "            \n",
    "        \n",
    "    # finds scientific names\n",
    "    def __find_scientific_names(self):\n",
    "        start = re.search(\"Assemblies used in these alignments:\", self.html)\n",
    "        end = re.search(\"---------------------------------------------------------------\", self.html)\n",
    "        text = self.html[start.end():end.start()]\n",
    "        names = []\n",
    "        for line in text.splitlines():\n",
    "            if (line != '') and not line.isspace() and line[0] != '=':\n",
    "                names.append(line[25:56].strip(' '))\n",
    "        self.scientific_names = names\n",
    "        return names\n",
    "\n",
    "\n",
    "    #finds names for alignment\n",
    "    def __find_alignment_names(self):\n",
    "        names = []\n",
    "        start = re.search(\"Assemblies used in these alignments:\", self.html)\n",
    "        end = re.search(\"---------------------------------------------------------------\", self.html)\n",
    "        text = self.html[start.end():end.start()]\n",
    "        for line in text.splitlines():\n",
    "            if (line == '') or line.isspace() or line[0] == '=':\n",
    "                continue\n",
    "            names.append(re.findall(\"/\\S*\\s\", line)[-1][1:-1])\n",
    "        self.alignment_names = names\n",
    "        return names\n",
    "    \n",
    "    #find common names\n",
    "    def __find_common_names(self):\n",
    "        start = re.search(\"Assemblies used in these alignments:\", self.html)\n",
    "        end = re.search(\"---------------------------------------------------------------\", self.html)\n",
    "        text = self.html[start.end():end.start()]\n",
    "        names = []\n",
    "        for line in text.splitlines():\n",
    "            if (line != '') and not line.isspace() and line[0] != '=':\n",
    "                names.append(line[:25].strip(' '))\n",
    "        self.common_names = names\n",
    "        return names\n",
    "    \n",
    "    #gets the alignment names for the 3 species given to it\n",
    "    def get_species_alignment_names(self, species = [\"Ovis aries\", \"Capra hircus\", \"Bos taurus\"]):\n",
    "        \n",
    "        if not hasattr(self, \"scientific_names\") or not hasattr(self, \"alignment_names\"):\n",
    "            self.common_names = self.__find_common_names()\n",
    "            self.scientific_names = self.__find_scientific_names()\n",
    "            self.alignment_names = self.__find_alignment_names()\n",
    "        \n",
    "        scientific_names = self.scientific_names\n",
    "        alignment_names = self.alignment_names\n",
    "        common_names = self.common_names\n",
    "        \n",
    "        species = self.species_names\n",
    "        chosen_species = [None]*3\n",
    "\n",
    "        # 3 chosen taxa = Ovis aries, Capra hircus and outgroup Bos taurus\n",
    "        chosen_species[0] = np.array(alignment_names)[(np.array(scientific_names) == species[0])][0]\n",
    "        chosen_species[1] = np.array(alignment_names)[(np.array(scientific_names) == species[1])][0]\n",
    "        chosen_species[2] = np.array(alignment_names)[(np.array(scientific_names) == species[2])][0]\n",
    "\n",
    "        self.chosen_species = chosen_species\n",
    "        return chosen_species"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "537905fa-d933-4f32-a3e5-91525cda52ea",
   "metadata": {},
   "source": [
    "All the chosen species triplets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "51302bc7-8b04-4221-bc59-8a5af1175d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chosen species for extraction\n",
    "common_species = [['Human', 'Chimp', 'Gorilla'],\n",
    "                  ['Chimp', 'Gorilla', 'Orangutan'],\n",
    "                  ['Gorilla', 'Orangutan', 'Gibbon'],\n",
    "                  ['Orangutan', 'Gibbon', 'Rhesus'],\n",
    "                  ['Rhesus', 'Crab-eating macaque', 'Baboon'],\n",
    "                  ['Baboon', 'Crab-eating macaque', 'Green monkey'],\n",
    "                  ['Baboon', 'Green monkey', 'Gibbon'],\n",
    "                  ['Mouse', 'Rat', 'Prairie vole'],\n",
    "                  ['Brush-tailed rat', 'Chinchilla', 'Guinea pig'],\n",
    "                  ['Chinese hamster', 'Golden hamster', 'Prairie vole'],\n",
    "                  ['Prairie vole', 'Rat', 'Lesser Egyptian jerboa'],\n",
    "                  ['Guinea pig', 'Naked mole-rat', 'Lesser Egyptian jerboa'],\n",
    "                  ['Pacific walrus', 'Weddell seal', 'Panda'],\n",
    "                  ['Panda', 'Ferret', 'Dog'],\n",
    "                  [\"David's myotis bat\", 'Microbat', 'Big brown bat'],\n",
    "                  ['Microbat', 'Big brown bat', 'Megabat'],\n",
    "                  ['Black flying-fox', 'Megabat', 'Big brown bat'],\n",
    "                  ['Sheep', 'Domestic goat', 'Cow'],\n",
    "                  ['Dolphin', 'Killer whale', 'Tibetan antelope'],\n",
    "                  ['Bactrian camel', 'Alpaca', 'Tibetan antelope'],\n",
    "                  ['Tasmanian devil', 'Wallaby', 'Opossum'],\n",
    "                  ['White-throated sparrow', 'Medium ground finch', 'Collared flycatcher'],\n",
    "                  ['Zebra finch', 'Medium ground finch', 'Collared flycatcher'],\n",
    "                  ['Parrot', 'Scarlet macaw', 'Budgerigar'],\n",
    "                  ['Peregrine falcon', 'Saker falcon', 'Budgerigar'],\n",
    "                  ['Chicken', 'Turkey', 'Mallard duck'],\n",
    "                  ['Chinese softshell turtle', 'Spiny softshell turtle', 'Green seaturtle'],\n",
    "                  ['Zebra mbuna', 'Pundamilia nyererei', \"Burton's mouthbreeder\"],\n",
    "                  ['Pundamilia nyererei', \"Burton's mouthbreeder\", 'Princess of Burundi'],\n",
    "                  [\"Burton's mouthbreeder\", 'Princess of Burundi', 'Nile tilapia'],\n",
    "                  ['Yellowbelly pufferfish', 'Fugu', 'Tetraodon']]\n",
    "            \n",
    "                  \n",
    "scientific_species = [['Homo sapiens', 'Pan troglodytes', 'Gorilla gorilla gorilla'],\n",
    "           ['Pan troglodytes', 'Gorilla gorilla gorilla', 'Pongo pygmaeus abelii'],\n",
    "           ['Gorilla gorilla gorilla', 'Pongo pygmaeus abelii', 'Nomascus leucogenys'],\n",
    "           ['Pongo pygmaeus abelii', 'Nomascus leucogenys', 'Macaca mulatta'],\n",
    "           ['Macaca mulatta', 'Macaca fascicularis', 'Papio hamadryas'],\n",
    "           ['Papio hamadryas', 'Macaca fascicularis', 'Chlorocebus sabaeus'],\n",
    "           ['Papio hamadryas', 'Chlorocebus sabaeus', 'Nomascus leucogenys'],\n",
    "           ['Mus musculus', 'Rattus norvegicus', 'Microtus ochrogaster'],\n",
    "           ['Octodon degus', 'Chinchilla lanigera', 'Cavia porcellus'],\n",
    "           ['Cricetulus griseus', 'Mesocricetus auratus', 'Microtus ochrogaster'],\n",
    "           ['Microtus ochrogaster', 'Rattus norvegicus', 'Jaculus jaculus'],\n",
    "           ['Cavia porcellus', 'Heterocephalus glaber', 'Jaculus jaculus'],\n",
    "           ['Odobenus rosmarus divergens', 'Leptonychotes weddellii', 'Ailuropoda melanoleuca'],\n",
    "           ['Ailuropoda melanoleuca', 'Mustela putorius furo', 'Canis lupus familiaris'],\n",
    "           ['Myotis davidii', 'Myotis lucifugus', 'Eptesicus fuscus'],\n",
    "           ['Myotis lucifugus', 'Eptesicus fuscus', 'Pteropus vampyrus'],\n",
    "           ['Pteropus alecto', 'Pteropus vampyrus', 'Eptesicus fuscus'],\n",
    "           ['Ovis aries', 'Capra hircus', 'Bos taurus'],\n",
    "           ['Tursiops truncatus', 'Orcinus orca', 'Pantholops hodgsonii'],\n",
    "           ['Camelus ferus', 'Vicugna pacos', 'Pantholops hodgsonii'],\n",
    "           ['Sarcophilus harrisii', 'Macropus eugenii', 'Monodelphis domestica'],\n",
    "           ['Zonotrichia albicollis', 'Geospiza fortis', 'Ficedula albicollis'],\n",
    "           ['Taeniopygia guttata', 'Geospiza fortis', 'Ficedula albicollis'],\n",
    "           ['Amazona vittata', 'Ara macao', 'Melopsittacus undulatus'],\n",
    "           ['Falco peregrinus', 'Falco cherrug', 'Melopsittacus undulatus'],\n",
    "           ['Gallus gallus', 'Meleagris gallopavo', 'Anas platyrhynchos'],\n",
    "           ['Pelodiscus sinensis', 'Apalone spinifera', 'Chelonia mydas'],\n",
    "           ['Maylandia zebra', 'Pundamilia nyererei', 'Haplochromis burtoni'],\n",
    "           ['Pundamilia nyererei', 'Haplochromis burtoni', 'Neolamprologus brichardi'],\n",
    "           ['Haplochromis burtoni', 'Neolamprologus brichardi', 'Oreochromis niloticus'],\n",
    "           ['Takifugu flavidus', 'Takifugu rubripes', 'Tetraodon nigroviridis']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12d1eced-8f0e-4061-a489-baeed1082e7f",
   "metadata": {},
   "source": [
    "Getting the common, scientific and alignment names for the species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "88b7d9c5-85cf-40d1-b2e2-72ef62f7336e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mp = MPMutExtractor()\n",
    "mp.get_species_alignment_names()\n",
    "scientific_names = np.array(mp.scientific_names)\n",
    "alignment_names = np.array(mp.alignment_names)\n",
    "common_names = np.array(mp.common_names)\n",
    "\n",
    "scientific_species = [[scientific_names[common_names == s][0] for s in species_triplet] for species_triplet in common_species]\n",
    "alignment_species = [[alignment_names[common_names == s][0] for s in species_triplet] for species_triplet in common_species]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88a62422-487f-43d7-a0c8-84fa37d3cf07",
   "metadata": {},
   "source": [
    "Creating the argumets file for condor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c2c73b1f-0586-4cc5-b323-b36d6071a74b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create input file for condor\n",
    "with open(\"species_arguments.txt\", \"w\") as f:\n",
    "    for i in range(len(alignment_species)):\n",
    "        f.write(f\"{';'.join(scientific_species[i]).replace(' ', '_')},{';'.join(alignment_species[i])}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
